### ISSUE TITLE
roundtrip impossible

### ISSUE THREAD

**Issue by psteinb** [LLNL/zfp#19]
Hi,

I am playing with zfp to compress some cryoEM data. I am struggling to perform a roundtrip, so I am doing the following with the current master:

```
/path/to/zfp -x omp=40 -f -i Tomo147_odd_topCrop.raw -o Tomo147_odd_topCrop_rt_a0.0001.raw -3 1092 3977 2580 -a 0.0001
```

I was hoping that the output file would match in size to the input, which it doesn't. The output is 4x smaller than the input. Which suprises me? The same happens if I do:

```
$ /path/to/zfp -i test.raw -z test.zfp <other options>
$ /path/to/zfp -z test.zfp -o roundtrip.raw <other options>
```
Am I overseeing something?
Best,
Peter

**Comment by lindstro** [comment]
Using /dev/zero as input, I can verify that zfp does produce a much smaller output file that is 10458824512 bytes long, which coincidentally is exactly 2^35 bytes = 34359738368 bytes = 2^32 words shorter than the 44818562880 bytes it should be.  I suspect that what should be a 64-bit index or counter somewhere gets truncated to 32 bits due to mixing 32-bit and 64-bit integers in the code.

We will look into this and provide a fix as soon as possible.  In the meantime, you might try compressing the data in smaller chunks by setting the z dimension (2580, in this case) to something smaller, e.g. 200, which evenly divides the number of threads you're using.  This ensures a good load balance and data chunks that are smaller than 4 GB = 2^32 bytes each.

**Comment by psteinb** [comment]
@lindstro thanks for looking into this. Indeed my input data is 40GB in size ... so passing the 4GB boundary is simple in this scenario. I am looking into different compression schemes currently. As we want to see if the downstream operations are affected by lossy zfp compression, it would be vital to have the fix you are mentioning. I'll try to bypass it for now.

**Comment by lindstro** [comment]
@psteinb, I'm fairly confident that it's not a file size issue but rather something to do with the number of array elements, which exceeds 2^32 in this case.  The issue could well be with the zfp executable and not the library itself.  Either way, this is a high priority issue for us that we're hoping to address soon.

**Comment by lindstro** [comment]
As suspected, the primary issue was with the zfp utility, which did not handle arrays with more than 2^32 elements properly.  We also addressed a related issue in the OpenMP compressor.

We have pushed a bug fix to the 'develop' branch.  @psteinb, can you please verify that this fix addresses your issue?

**Comment by psteinb** [comment]
it works perfectly now at least by checking file sizes. Thanks for providing the fix. Feel free to close this issue, once the fix is in `master`.

**Comment by lindstro** [comment]
Thanks for verifying.

**Comment by lindstro** [comment]
The fix is now in master and the 0.5.4 release.
